analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Nullable.cs

```diff
@@ -59,4 +59,30 @@ public void Nullable_Value_ReadsConstraintsFromInstance()
         validator.ValidateTag("FalseFirst", x => x.HasConstraint(BoolConstraint.False).Should().BeTrue());
         validator.ValidateTag("FalseFirst", x => x.HasConstraint(TestConstraint.First).Should().BeTrue());
     }
+
+    [TestMethod]
+    public void Nullable_HasValue_ReadsBoolConstraintFromObjectConstraint()
+    {
+        const string code = """
+            var hasValue = arg.HasValue;
+            Tag("HasValueUnknown", hasValue);
+            Tag("SymbolUnknown", arg);
+            arg = true;
+            hasValue = arg.HasValue;
+            Tag("HasValueAfterTrue", hasValue);
+            Tag("SymbolAfterTrue", arg);
+            arg = null;
+            hasValue = arg.HasValue;
+            Tag("HasValueAfterNull", hasValue);
+            Tag("SymbolAfterNull", arg);
+            """;
+        var validator = SETestContext.CreateCS(code, ", bool? arg").Validator;
```

Comment:
The test would be easier to read if we use `int?` instead. It's confusing to keep track of `HasValue`, `arg`, and `BoolConstraint` at the same time. Maybe add a second test case with `int?`?

Reply:
`int` doesn't have any constraint on it. I wouldn't mind if you rewrite it in your PRs. As there's ` ToDo` for you here anyway :)